### ISSUE TITLE
Make `general_mat_mul`/`general_mat_vec_mul` more discoverable

### ISSUE THREAD

**Issue by jturner314** [rust-ndarray/ndarray#927]
The [`general_mat_mul`](https://docs.rs/ndarray/0.14.0/ndarray/linalg/fn.general_mat_mul.html) and [`general_mat_vec_mul`](https://docs.rs/ndarray/0.14.0/ndarray/linalg/fn.general_mat_vec_mul.html) functions are not easy to find, since they're in the `linalg` module while most other functionality is available either at the top level of the crate or as methods on the array types.

At a minimum, we should mention them in the docs for the `dot` methods and `Dot` trait.

However, I think it would also be useful to expose the functionality as methods on `ArrayBase`. I'd suggest adding a trait like this:

```rust
/// General matrix multiplication.
pub trait GeneralDot<A, Rhs1, Rhs2> {
    /// Perform general matrix-matrix multiplication, modifying `self` in-place.
    ///
    /// Compute C ← α A B + β C, where `self` is C.
    ///
    /// The array shapes must agree in the way that if `a` is *M* × *N*, then
    /// `b` is *N* × *K* and `self` is *M* × *K*.
    ///
    /// ***Panics*** if array shapes are not compatible<br>
    /// *Note:* If enabled, uses blas `gemm` for elements of `f32, f64` when memory
    /// layout allows.  The default matrixmultiply backend is otherwise used for
    /// `f32, f64` for all memory layouts.
    fn mul_add_scaled_dot_inplace(
        &mut self,
        beta: A,
        alpha: A,
        a: &Rhs1,
        b: &Rhs2,
    );
}
```

and then add `mul_add_scaled_dot_inplace` methods which use this trait to `ArrayBase<S, Ix1>` and `ArrayBase<S, Ix2>`. What do you think?

**Comment by bluss** [comment]
The method name `mul_add_scaled_dot_inplace` is rather inscrutable but yes, it would be great to make it easier to discover. How do we know if A, B or C should be self? C fits but I think users would understand it even better if it was A?

I think we'll have some more "operators" being added as free functions - things similar to the `assign_to` free function that now lives in https://github.com/rust-ndarray/ndarray/blob/f7a12775d4ee5fd007c58a35cdf84b3b9c289c45/src/traversal_utils.rs

The reason is that free functions are easier to use with generics of the kind that `assign_to` needs (no certain Self type) - and it might be interesting in the future to expose functions like that.

**Comment by jturner314** [comment]
Yeah, I agree the name `mul_add_scaled_dot_inplace` is hard to read. Maybe `mul_add_assign_scaled_dot` would be better? I think that in most cases, it makes sense for the array being mutated to be `self`.

We could have a few methods, of increasing complexity:

1. `assign_dot`: `self = AB`
2. `add_assign_dot`: `self += AB`
3. `add_assign_scaled_dot`: `self += alpha*AB`
4. `mul_add_assign_scaled_dot`: `self = beta*self + alpha*AB`

Or, maybe some subset of these (e.g. 1 and 2, or 1, 2, and 3), and leave the most general case to a standalone function.

I wonder if moving stuff out of the top level and into modules would help with discoverability, just because users would be more likely to look at the modules.

I do think that mentioning standalone functions in the relevant places in the `ArrayBase` docs would be sufficient.

**Comment by bluss** [comment]
I'd rather give things memorable names; the systematic names like that don't seem to help

**Comment by bluss** [comment]
Maybe something like this for `alpha A B + beta C -> C`: `a.mat_mul_scaled_into(alpha, &b, beta, &mut c)`

**Comment by jturner314** [comment]
Hmm... The name `mat_mul_scaled_into` makes me think the operation is `alpha A B -> C`, although the presence of the `beta` parameter does make that interpretation less likely.

Maybe we should just leave the general case as a standalone function and mention it in the docs of the methods. In that case, I think it would be useful to have at least `assign_dot` and `add_assign_dot` methods (or `assign_scaled_dot` and `add_assign_scaled_dot`, or if we want `a` to be `self`, something like `dot_assign_to` and `dot_add_assign_to`).
